third_party/amd/lib/Analysis/RangeAnalysis.cpp

```diff
@@ -60,45 +62,142 @@ void getEnclosingLoops(Operation &op, SmallVector<LoopLikeOpInterface> &ops) {
+  assert(op->getNumOperands() == 2 && "expected op to have two operands");
+  assert(argRanges.size() == 2 && "expected two arg ranges");
+  setResultRange(op->getResult(0), argRanges[0].rangeUnion(argRanges[1]));
+}
+
+void inferResultRangesMaxNonNegSigned(Operation *op,
+                                      SetIntRangeFn setResultRange) {
+  for (auto result : op->getResults()) {
+    auto bitWidth =
+        mlir::ConstantIntRanges::getStorageBitwidth(result.getType());
+    setResultRange(result, ConstantIntRanges::fromSigned(
+                               APInt::getZero(bitWidth).sext(bitWidth),
+                               APInt::getMaxValue(bitWidth).sext(bitWidth)));
+  }
 }
 
 } // namespace
 
 namespace mlir::triton::AMD {
 
+bool ConstantRangeIsStaticallyTrue(const ConstantIntRanges &range, Type type) {
+  auto bitWidth = mlir::ConstantIntRanges::getStorageBitwidth(type);
+  if (bitWidth != 1)
+    return false;
+  if (auto maybeCVal = range.getConstantValue();
+      maybeCVal.value_or(APInt::getZero(bitWidth)) ==
+      APInt::getAllOnes(bitWidth))
+    return true;
+  return false;
+}
+
 void TritonIntegerRangeAnalysis::setToEntryState(
     dataflow::IntegerValueRangeLattice *lattice) {
-  propagateIfChanged(lattice, lattice->join(IntegerValueRange::getMaxRange(
-                                  lattice->getAnchor())));
+  auto anchor = lattice->getAnchor();
+  auto range = IntegerValueRange::getMaxRange(anchor);
+  auto maybeCmpOpUser = llvm::find_if(anchor.getUsers(), [&](Operation *op) {
+    return op->getNumResults() == 1 && assumptions.contains(op->getResult(0));
+  });
+  if (maybeCmpOpUser != anchor.getUsers().end()) {
+    auto cmpOp = llvm::cast<arith::CmpIOp>(*maybeCmpOpUser);
+    bool anchorIsLhs = cmpOp.getLhs() == anchor;
+    auto maybeConstantIntValue = getConstantIntValue(
+        getAsOpFoldResult(anchorIsLhs ? cmpOp.getRhs() : cmpOp.getLhs()));
+    if (auto constValue = maybeConstantIntValue; constValue.value_or(-1) > 0) {
+      auto operTy = llvm::cast<IntegerType>(anchor.getType());
+      unsigned bitWidth = operTy.getWidth();
+      bool isSigned = operTy.isSigned();
+      APInt apVal = {bitWidth, static_cast<uint64_t>(*constValue), isSigned};
+      APInt min = APInt::getMinValue(bitWidth);
+      APInt max = APInt::getMaxValue(bitWidth);
+
+      switch (cmpOp.getPredicate()) {
+      case arith::CmpIPredicate::eq:
+        range = mlir::ConstantIntRanges::constant(apVal);
+        break;
+      case arith::CmpIPredicate::sge:
+        if (anchorIsLhs)
+          range = mlir::ConstantIntRanges::range(apVal, max, isSigned);
+        else
+          range = mlir::ConstantIntRanges::range(min, apVal, isSigned);
+        break;
+      case arith::CmpIPredicate::sgt:
+        if (anchorIsLhs)
+          range = mlir::ConstantIntRanges::range(apVal + 1, max, isSigned);
+        else
+          range = mlir::ConstantIntRanges::range(min, apVal + 1, isSigned);
```

Comment:
Is this correct? Shouldn't it be `(min, appVal - 1)`?

Can you add tests for this case?

Reply:
```
apVal > K
=>
apVal - 1 >= K
```

Whoops must've been asleep when I wrote that - ya will add test cases here and for the case below.

Reply:
fixed